sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx.tp

```diff
@@ -0,0 +1,220 @@
+
+cpdef enum WeightingStrategy:
+    uniform = 0
+    distance = 1
+    other = 2
+
+{{for name_suffix, INPUT_DTYPE_t, INPUT_DTYPE in implementation_specific_values}}
+from ._argkmin cimport ArgKmin{{name_suffix}}, EuclideanArgKmin{{name_suffix}}
+from ._datasets_pair cimport DatasetsPair{{name_suffix}}
+
+cdef class ArgKminLabels{{name_suffix}}(ArgKmin{{name_suffix}}):
+    """
+    {{name_suffix}}bit implementation of ArgKminLabel.
+    """
+    cdef:
+        const ITYPE_t[:] labels,
+        DTYPE_t[:, :] label_weights
+        cpp_map[ITYPE_t, ITYPE_t] labels_to_index
+        WeightingStrategy weight_type
+
+    @classmethod
+    def compute(
+        cls,
+        X,
+        Y,
+        ITYPE_t k,
+        weights,
+        labels,
+        str metric="euclidean",
+        chunk_size=None,
+        dict metric_kwargs=None,
+        str strategy=None,
+    ):
+        """Compute the argkmin reduction with labels.
+
+        This classmethod is responsible for introspecting the arguments
+        values to dispatch to the most appropriate implementation of
+        :class:`ArgKminLabels{{name_suffix}}`.
+
+        This allows decoupling the API entirely from the implementation details
+        whilst maintaining RAII: all temporarily allocated datastructures necessary
+        for the concrete implementation are therefore freed when this classmethod
+        returns.
+
+        No instance _must_ directly be created outside of this class method.
+        """
+        # Use a generic implementation that handles most scipy
+        # metrics by computing the distances between 2 vectors at a time.
+        pda = ArgKminLabels{{name_suffix}}(
+            datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+            k=k,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            weights=weights,
+            labels=labels,
+        )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pda.execute_in_parallel_on_Y:
+                pda._parallel_on_Y()
+            else:
+                pda._parallel_on_X()
+
+        return pda._finalize_results()
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        const ITYPE_t[:] labels,
+        chunk_size=None,
+        strategy=None,
+        ITYPE_t k=1,
+        weights=None,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            k=k,
+        )
+
+        if weights == "uniform":
+            self.weight_type = WeightingStrategy.uniform
+        elif weights == "distance":
+            self.weight_type = WeightingStrategy.distance
+        else:
+            self.weight_type = WeightingStrategy.other
```

Comment:
It's not clear where this `WeightingStrategy.other` is used. I cannot find it referenced in the diff of the PR. If it's never used, let's remove it.

In the future, it might make sense to have `WeightingStrategy.custom` or `WeightingStrategy.callable` option to make it possible to pass a custom Cython (or even Python?) function/callable to compute the weights from the distances?

Such an option would be very useful, in particular for the regression equivalent of this reducer to be able to implement some kind of LOWESS regressor in scikit-learn as discussed in this SO reply: https://stackoverflow.com/a/44915533

But passing custom distance weighting/aggregation functions, especially with user settable parameters is definitely beyond the scope of this PR. It was just food for thought.

Reply:
Indeed the original intention was to offer it as an encoding for callables, hence I've changed `WeightingStrategy.other --> WeightingStrategy.callable` and introduced a couple `TODO` comments explaining that only `"uniform"` is implemented right now. I think it would be good to keep it as-is but am open to removing it for now if you feel it is appropriate.

Reply:
Alright. Not sure if it's a YAGNI or not but the extra complexity is not large and the TODO comments are helpful to explain the intent for the future of this code base.